feat(helm): Base Helm chart for Sefaria project.

parth-gohil opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/parth-gohil/sefaria-project/1)
<!-- Reviewable:end -->

parth-gohil committed:
feat(helm): Initial helm chart from k8s-admin repo

parth-gohil committed:
fix(helm): Removes unnecessary resource values.

parth-gohil committed:
fix(helm): Breaks down the image values futher for better templating.

parth-gohil committed:
fix(helm): Breaks down the Varnish image values futher for better templating.

parth-gohil committed:
fix(helm): Removes the burden from values file to add '/' in values

parth-gohil committed:
fix(helm): Updates image name templating for nginx rollout.

parth-gohil committed:
fix(helm): Updates image name templating for the monitor rollout.

moshloop commented on helm-chart/scripts/mongo/create-dumps.sh:
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env bash
```
I think the mongo backup, restore, upload jobs can come out for now

moshloop commented on helm-chart/templates/cronjobs/regenerate_long_cached_data.yaml:
```diff
@@ -0,0 +1,41 @@
+{{if eq (substr 0 4 .Values.deployEnv) "prod"}}
```
I think we should make this `eq .values.DeployEnv "prod"` across the board

moshloop commented on helm-chart/templates/cronjobs/sitemaps.yaml:
```diff
@@ -0,0 +1,63 @@
+{{if eq (substr 0 4 .Values.deployEnv) "prod"}}
```
I think this can come out as its not a core part of the app

parth-gohil replied:
Should I be removing the condition or the whole cronjob?

moshloop commented on helm-chart/templates/ingress.yaml:
```diff
@@ -0,0 +1,114 @@
+{{if eq (substr 0 4 .Values.deployEnv) "prod"}}
+---
+apiVersion: extensions/v1beta1
```
we should change theses to networking/v1

moshloop commented on helm-chart/templates/ingress.yaml:
```diff
@@ -0,0 +1,114 @@
+    - secretName: origin-tls-{{ .Values.deployEnv }}
+    - secretName: origin-il-tls-{{ .Values.deployEnv }}
+  rules:
+    - host: "{{ substr 0 4 .Values.deployEnv }}.sefaria.org"
```
this should be a range over `.Values.ingress.hosts`

moshloop commented on helm-chart/templates/secrets.yaml:
```diff
@@ -0,0 +1,106 @@
+    deployEnv: "{{ .Values.deployEnv }}"
+type: Opaque
+data:
+  BackupManagerKey.json: {{ .Files.Get "secrets/BackupManagerKey.json" | b64enc }}
```
the filenames should be variables in the values.yml

BrendanGalloway replied:
Do we want it to be a file reference variable? The solution we discussed previously was to provide a ref to an existing secret, with an option to pass in data to create a new secret. That might not align with the proposal to manage cauldrons via gitops, since there isn't really a way to encrypt the values section of a helmRelease that is stored in repo. A file reference var definitely won't work with a helmRelease

moshloop replied:
👍

moshloop commented on helm-chart/templates/secrets.yaml:
```diff
@@ -0,0 +1,106 @@
+    deployEnv: "{{ .Values.deployEnv }}"
+type: Opaque
+data:
+  slack-webhook: {{ "https://hooks.slack.com/services/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" | b64enc | quote }}
```
Change to variable,

moshloop commented on helm-chart/values.yaml:
```diff
@@ -0,0 +1,108 @@
+      cpu: "1500m"
+
+varnish:
+  containerImage:
```
I would change this a string of the full container image repo/path/version - The calculation of the image tag should be external to the helm chart.

BrendanGalloway replied:
Separate repo/path/image fields are fairly standard though, since it lets a consumer of the chart point at a mirrored repo while still consuming bumps to default image numbers through releases.

moshloop replied:
Perhaps  - but then there is no need for a separate repoLocation and repoName it would should be container.image and container.version

BrendanGalloway reviewed (changes requested):
At a high level it's mostly just secret referencing that needs to be updated, and then maybe a little bit of layout cleanup

> Reviewed 31 of 31 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-:-MuL-rVL8vdFKGG85oNB:bg2a92a)* status: all files reviewed, 9 unresolved discussions (waiting on @parth-gohil)

___
*[.gitignore, line 133 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKzmYo7UQsDO3M_LvN:-MuKzmYpC65BDccJFUoB:b77xgz7) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/.gitignore#L133)):*
> ```ignore
> #################
> /dafChat/deploy/dafchat/valuesGeneration/_generatedHelmValues.yaml
> helm-chart/pipeline-flow.md
> ```

Why are we excluding this from git?
___
*[helm-chart/gunicorn.conf.py, line 1 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKzx2s457PTYZo9m-d:-MuKzx2s457PTYZo9m-e:bpb7qlo) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/gunicorn.conf.py#L1)):*
> ```Python
> # args: [ "python manage.py migrate && gunicorn sefaria.wsgi --access-logfile - --error-logfile /log/gunicorn-error.log --timeout 300 --worker-class=gevent --worker-connection 2000 --worker-tmp-dir /dev/shm -b 0.0.0.0:80" ]
> ```

Where is this file being used?  it should probably not sit in the root directory
___
*[helm-chart/nginx.template.conf, line 1 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKfANq3-p22FoGOqzo:-MuKfANq3-p22FoGOqzp:b-9zzk3q) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/nginx.template.conf#L1)):*
> ```
> user www-data;
> ```

where is this file consumed?  Probably shouldn't be in the root directory.
___
*[helm-chart/README.md, line 0 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKWeCu7X0M9A5kRkQi:-MuKWeCu7X0M9A5kRkQj:b-uaxv4k) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/b086e3f7ab4944602b18f421d51f5f794a6fb840/helm-chart/README.md#L0)):*
Was this file empty in the original? Either delete or populate.
___
*[helm-chart/values.yaml, line 1 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKXHNRCY-tQnI770gR:-MuKXHNRCY-tQnI770gS:b-33h53) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/values.yaml#L1)):*
> ```YAML
> ---
> ```

I would add some additional commentary here describing what some of the options do in additional to info about restrictions/how they should be populated.  At high level deployEnv determines the naming of the deployed set of pods and associated congifmaps etc, sandbox true/false sets x, y and z in the system etc.
___
*[helm-chart/_assets/placeholder, line 1 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKfORJ419fDHFF9vVO:-MuKfORJ419fDHFF9vVP:bh11lzj) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/_assets/placeholder#L1)):*
> ```
> placeholder
> ```

what is this holding place for?
___
*[helm-chart/scripts/mongo/create-dumps-no-history.sh, line 0 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuL-4BU8zmTCuFP_8bj:-MuL-4BU8zmTCuFP_8bk:b-jw9d91) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/b086e3f7ab4944602b18f421d51f5f794a6fb840/helm-chart/scripts/mongo/create-dumps-no-history.sh#L0)):*
Is this being used anywhere?  Can probably delete if not
___
*[helm-chart/templates/secrets.yaml, line 1 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKYvTW8v8di9RO8Vv2:-MuKYvTW8v8di9RO8Vv3:b-z2kyc0) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/templates/secrets.yaml#L1)):*
> ```YAML
> ---
> ```

This file needs to be reworked since the referenced objects are no longer in repo.  

wrap each secret with a test to determine if the secret contents are being passed in via values, and then format name and data accordingly.
___
*[helm-chart/templates/cronjobs/mongobackup.yaml, line 87 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKgQgeCCUVigDl02-E:-MuKgQgeCCUVigDl02-F:bhv2n8w) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/templates/cronjobs/mongobackup.yaml#L87)):*
> ```YAML
> ---
> apiVersion: v1
> kind: Secret
> ```

This secret and the next can probably be configmaps?  I did not see anything sensitive in the scripts themselves apart from maybe the bucket name?

TODO: make the scripts more configurable, with ENVVAR references to options like destination bucket that can be populated by an EnvFrom configmap/secret, eg for operation in forked projects


<!-- Sent from Reviewable.io -->

parth-gohil committed:
fix(helm): Removes the mongo scripts.

parth-gohil committed:
fix(helm): Updates the conditions to validate with prod.

parth-gohil committed:
fix(helm): Updates to latest API version for Ingress object

parth-gohil committed:
fix(helm): Templatizes ingress hosts values.

parth-gohil committed:
fix(helm): Slack Webhook value templatizing.

parth-gohil reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/parth-gohil/sefaria-project/1)* status: all files reviewed, 17 unresolved discussions (waiting on @BrendanGalloway, @moshloop, and @parth-gohil)

___
*[.gitignore, line 133 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKzmYo7UQsDO3M_LvN:-MuPEFHU3YWw0C0hCSJp:be3re5b) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/.gitignore#L133)):*
<details><summary><i>Previously, BrendanGalloway (Brendan Galloway) wrote…</i></summary><blockquote>

Why are we excluding this from git?
</blockquote></details>

This was a local file created to understand the flow of the pipeline. I'll remove this from the .gitignore. 
___
*[helm-chart/gunicorn.conf.py, line 1 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKzx2s457PTYZo9m-d:-MuPT7TC2Nm8lvP6PADu:bwu7lvu) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/gunicorn.conf.py#L1)):*
<details><summary><i>Previously, BrendanGalloway (Brendan Galloway) wrote…</i></summary><blockquote>

Where is this file being used?  it should probably not sit in the root directory
</blockquote></details>

Gets used in the web rollout as a secret. Will move such to the _assets directory.
___
*[helm-chart/nginx.template.conf, line 1 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKfANq3-p22FoGOqzo:-MuPTpE_8Km8jfbsguyv:b-e4eonx) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/nginx.template.conf#L1)):*
<details><summary><i>Previously, BrendanGalloway (Brendan Galloway) wrote…</i></summary><blockquote>

where is this file consumed?  Probably shouldn't be in the root directory.
</blockquote></details>

Used in the nginx rollout as a secret mount. Moved it to _assets directory.
___
*[helm-chart/README.md, line  at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKWeCu7X0M9A5kRkQi:-MuPTwky35x4dghD_VkG:b-okdqy9) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/b086e3f7ab4944602b18f421d51f5f794a6fb840/helm-chart/README.md#L)):*
<details><summary><i>Previously, BrendanGalloway (Brendan Galloway) wrote…</i></summary><blockquote>

Was this file empty in the original? Either delete or populate.
</blockquote></details>

Deleting for now. Will populate with fresh instructions once we have frozen the chart.
___
*[helm-chart/values.yaml, line 1 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKXHNRCY-tQnI770gR:-MuPUDnXCZ5bVV1CSNQE:b-g10nb1) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/values.yaml#L1)):*
<details><summary><i>Previously, BrendanGalloway (Brendan Galloway) wrote…</i></summary><blockquote>

I would add some additional commentary here describing what some of the options do in additional to info about restrictions/how they should be populated.  At high level deployEnv determines the naming of the deployed set of pods and associated congifmaps etc, sandbox true/false sets x, y and z in the system etc.
</blockquote></details>

Do you reckon we can populate the documentation for this in the README.md file? 
___
*[helm-chart/_assets/placeholder, line 1 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKfORJ419fDHFF9vVO:-MuPUPD34kZondrRB6t_:b-v6s1o3) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/_assets/placeholder#L1)):*
<details><summary><i>Previously, BrendanGalloway (Brendan Galloway) wrote…</i></summary><blockquote>

what is this holding place for?
</blockquote></details>

Removed. 
___
*[helm-chart/scripts/mongo/create-dumps-no-history.sh, line  at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuL-4BU8zmTCuFP_8bj:-MuPUWNd7VHOZ3cm05PG:b-ksr0pg) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/b086e3f7ab4944602b18f421d51f5f794a6fb840/helm-chart/scripts/mongo/create-dumps-no-history.sh#L)):*
<details><summary><i>Previously, BrendanGalloway (Brendan Galloway) wrote…</i></summary><blockquote>

Is this being used anywhere?  Can probably delete if not
</blockquote></details>

Removed this file. Was also asked by Moshe to remove the scripts out of helm for now. 
___
*[helm-chart/templates/ingress.yaml, line 3 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKWQSL2Hjdc7F3P-Qd-r1-3:-MuPUiTBC5EOZ6vUNbPN:b-896fix) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/templates/ingress.yaml#L3)):*
<details><summary><i>Previously, moshloop wrote…</i></summary><blockquote>

we should change theses to networking/v1
</blockquote></details>

Done.
___
*[helm-chart/templates/ingress.yaml, line 14 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKWQSL2Hjdc7F3P-Qd-r1-14:-MuPUjvqBqU0hNcXNaHA:b-896fix) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/templates/ingress.yaml#L14)):*
<details><summary><i>Previously, moshloop wrote…</i></summary><blockquote>

this should be a range over `.Values.ingress.hosts`
</blockquote></details>

Done.
___
*[helm-chart/templates/secrets.yaml, line 85 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKWQSL2Hjdc7F3P-Ql-r1-85:-MuPV4O-60n5k00r5Hak:b-896fix) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/templates/secrets.yaml#L85)):*
<details><summary><i>Previously, moshloop wrote…</i></summary><blockquote>

Change to variable, 
</blockquote></details>

Done.
___
*[helm-chart/templates/cronjobs/mongobackup.yaml, line 87 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKgQgeCCUVigDl02-E:-MuPVM0S5zI6Xz6KCXSA:b4v03ud) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/templates/cronjobs/mongobackup.yaml#L87)):*
<details><summary><i>Previously, BrendanGalloway (Brendan Galloway) wrote…</i></summary><blockquote>

This secret and the next can probably be configmaps?  I did not see anything sensitive in the scripts themselves apart from maybe the bucket name?

TODO: make the scripts more configurable, with ENVVAR references to options like destination bucket that can be populated by an EnvFrom configmap/secret, eg for operation in forked projects
</blockquote></details>

Not sure what to do now since we've moved the scripts out of the repo as suggested by Moshe.
___
*[helm-chart/templates/cronjobs/regenerate_long_cached_data.yaml, line 1 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKWQSL2Hjdc7F3P-QZ-r1-1:-MuPVXi_77IZo4Y5dwlS:b-896fix) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/templates/cronjobs/regenerate_long_cached_data.yaml#L1)):*
<details><summary><i>Previously, moshloop wrote…</i></summary><blockquote>

I think we should make this `eq .values.DeployEnv "prod"` across the board
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

parth-gohil committed:
fix(helm): Moved file for better organisation.

parth-gohil committed:
fix(helm): Removed local files from gitignore.

moshloop commented on helm-chart/values.yaml:
```diff
@@ -0,0 +1,122 @@
+  replicaCount: 1
+  resources:
+    request:
+      memory: "16Gi" # must be in megabibytes, because we pass it into the malloc definition
```
These should be templated out, with more dev friendly defaults

BrendanGalloway reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/parth-gohil/sefaria-project/1)* status: 9 of 31 files reviewed, 14 unresolved discussions (waiting on @BrendanGalloway, @moshloop, and @parth-gohil)

___
*[helm-chart/values.yaml, line 1 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKXHNRCY-tQnI770gR:-MuQEYtj1ONgVwepYDyH:b-m61waa) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/values.yaml#L1)):*
<details><summary><i>Previously, parth-gohil (Parth Gohil) wrote…</i></summary><blockquote>

Do you reckon we can populate the documentation for this in the README.md file? 
</blockquote></details>

A quick and dirty workflow while setting up helm is to use `helm get values` to pull the default values file in order to look up what values can be set.  For this reason some projects use the values.yaml as a pseudo readme, or at least provide a bit of documentation in the values file itself.  See https://github.com/ondat/charts/blob/main/charts/ondat-operator/values.yaml vs https://github.com/ondat/charts/blob/main/charts/ondat-operator/README.md as an example
___
*[helm-chart/templates/cronjobs/mongobackup.yaml, line 87 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKgQgeCCUVigDl02-E:-MuQFZnf4DihVde2IAs4:b-5a7j3j) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/templates/cronjobs/mongobackup.yaml#L87)):*
<details><summary><i>Previously, parth-gohil (Parth Gohil) wrote…</i></summary><blockquote>

Not sure what to do now since we've moved the scripts out of the repo as suggested by Moshe.
</blockquote></details>

Ignore for now - I imagine at some point we will need to reintroduce these scripts either here or via infrastructure, at which point we can circle back to how these jobs are configured
___
*[helm-chart/gunicorn.conf.py, line 1 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKzx2s457PTYZo9m-d:-MuQHGItDz5Y5fm-i4w6:bpudg5u) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/gunicorn.conf.py#L1)):*
<details><summary><i>Previously, parth-gohil (Parth Gohil) wrote…</i></summary><blockquote>

Gets used in the web rollout as a secret. Will move such to the _assets directory.
</blockquote></details>

Does it need to be a secret?  If we're keeping it in plaintext it can probably just be a configmap
___
*[helm-chart/nginx.template.conf, line 1 at r1](https://reviewable.io/reviews/parth-gohil/sefaria-project/1#-MuKfANq3-p22FoGOqzo:-MuQHO6L2s-KEk8-k_GT:b-krvclc) ([raw file](https://github.com/parth-gohil/sefaria-project/blob/7bf5ce82d54393803e22d9819ec77ee72f515c8a/helm-chart/nginx.template.conf#L1)):*
<details><summary><i>Previously, parth-gohil (Parth Gohil) wrote…</i></summary><blockquote>

Used in the nginx rollout as a secret mount. Moved it to _assets directory.
</blockquote></details>

Can probably also change this to be a configmap


<!-- Sent from Reviewable.io -->

parth-gohil committed:
fix(helm): Documentation for some base configuration values.

parth-gohil committed:
fix(helm): Uses configmap to mount configuration files.

moshloop commented on helm-chart/values.yaml:
```diff
@@ -1,11 +1,28 @@
+
+# This is the project ID for images coming from gcr. Ther are two possible values for this currently development-205018 / production-deployment.
+# Should pass ${PROJECT_ID} value to this when applying the helm chart in the CI pipeline.
+projectID: development-205018 
```
This shouldn't be defaulted

moshloop replied:
and in fact we shouldn't need it at all, we should have a single variable called `imageRegistry` which would be the full path to either the gcr project or docker etc..

parth-gohil replied:
I am just wondering how we'd handle the different projectIDs that the images are being release under. Do we expect people to change the value for `imageRegistry` for their helm release? Is it understood that everyone using the helm will have the visibility of the available or possible registry names?

moshloop replied:
There are 2 use cases - the general public who will not have access to GCP - only to a public registry e.g. dockerhub or github packages, and the other to people who have access GCP. 

The default should be publicly available, and since there are none at this point, it should default to empty - We also shouldn't be specifying the project id at all, rather just the imageRegistry prefix

parth-gohil replied:
Noted.

moshloop commented on helm-chart/values.yaml:
```diff
@@ -1,11 +1,28 @@
+projectID: development-205018 
+
+# This is used in mongobackup & reindex-elasticsearch cron jobs.
 slackWebhook: "https://hooks.slack.com/services/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"
```
This is a secret value and should not be defaulted

parth-gohil replied:
I've removed the default from the values file.

moshloop commented on helm-chart/values.yaml:
```diff
@@ -1,11 +1,28 @@
+# Need to pass $BUILD_ID to this key when applying the helm chart in CI. 
+# or 
+# Set the commit hash for a particular buiild to deploy built image for that commit ID. (NOTE: This would work only if the image is built for that commit ID.)
+releaseImageTag: cf66cfb 
```
this shouldn't be defaulted, or default to latest if they have those tags

BrendanGalloway replied:
If we want to manage a distribution of releases to consumers, then it should be defaulted to the tagged images associated with that release?

parth-gohil replied:
Removed the default value out of the values file for now. Will update the manifests to use `imageRegistry` and `tag` as two separate inputs for the image.

parth-gohil committed:
fix(helm): Keys for configmap name fixed.

parth-gohil committed:
fix(helm): Post templating white space removal.

parth-gohil committed:
fix(helm): Removes unnecessary defaults.

BrendanGalloway commented on helm-chart/templates/ingress.yaml:
```diff
@@ -0,0 +1,73 @@
+  tls.key: {{ .Files.Get (printf "secrets/tls/%s/origin-il.key" (substr 0 4 .Values.deployEnv)) | b64enc }}
+...
+{{end}}
+
```
This needs to be reworked as per the general approach for secret handling

BrendanGalloway commented on helm-chart/templates/ingress.yaml:
```diff
@@ -0,0 +1,73 @@
+spec:
+  tls:
+    - secretName: origin-tls-{{ .Values.deployEnv }}
+    - secretName: origin-il-tls-{{ .Values.deployEnv }}
```
These need to configurable as reference to preexisting objects

BrendanGalloway commented on helm-chart/templates/ingress.yaml:
```diff
@@ -0,0 +1,73 @@
+type: kubernetes.io/tls
+data:
+  tls.crt: {{ .Files.Get (printf "secrets/tls/%s/origin.crt" (substr 0 4 .Values.deployEnv)) | b64enc }}
+  tls.key: {{ .Files.Get (printf "secrets/tls/%s/origin.key" (substr 0 4 .Values.deployEnv)) | b64enc }}
```
Same action as the prod ingress

parth-gohil committed:
fix(helm): Apply secrets conditionally.

parth-gohil committed:
fix(helm): Setting Developer friendly defaults for resources.

parth-gohil committed:
fix(helm): Setting simpler helm templating for images key on all manifests.

BrendanGalloway commented on helm-chart/templates/configMaps.yaml:
```diff
@@ -0,0 +1,351 @@
+
+        location /static/sitemaps/ {
+          access_log off;
+          proxy_pass https://storage.googleapis.com/sefaria-sitemaps$request_uri;
```
This should be configurable

parth-gohil replied:
I somehow missed these comments about making the nginx config more helm friendly. 

There are two ways to do this. 
- The current implementation uses `envsubst` and replaces values in the file to generate a static config file and apply it in the nginx pod.
- The other way is we straight up use helm variables inside the ConfigMap string and configure from there. This way we won't have to mount additional ConfigMaps as environemnt variables and won't have to edit this `envsubst` command 
`[ "envsubst '${ENV_NAME},${VARNISH_HOST},${SEARCH_HOST},${RELEASE_TAG}' < /conf/nginx.template.conf > /nginx.conf && exec nginx -c /nginx.conf -g 'daemon off;'" ]`

Let me know which method makes more sense.

BrendanGalloway commented on helm-chart/templates/configMaps.yaml:
```diff
@@ -0,0 +1,351 @@
+      server {
+        listen 80;
+        listen [::]:80;
+        server_name sefaria.org sefaria.org.il;
```
these should be configurable (maybe read from ingress info?)

BrendanGalloway commented on helm-chart/templates/configMaps.yaml:
```diff
@@ -0,0 +1,351 @@
+          access_log off;
+          autoindex on;
+          default_type application/json;
+          return 200 '{"applinks": {"apps": [], "details": [{"appID": "2626EW4BML.org.sefaria.sefariaApp", "paths": ["*"]}]}}';
```
not sure if this appID needs to be configureable

BrendanGalloway commented on helm-chart/templates/cronjobs/mongobackup.yaml:
```diff
@@ -0,0 +1,106 @@
+  namespace: default
+data:
+  upload-dumps.sh: {{ .Files.Get "scripts/mongo/upload-dumps.sh" | b64enc }}
+...
```
Rework file references

parth-gohil committed:
fix(helm): Removes local files from .gitignore.

BrendanGalloway commented on helm-chart/templates/cronjobs/mongobackup.yaml:
```diff
@@ -0,0 +1,106 @@
+              secretName: backup-manager-secret
+          - name: create-dumps-script
+            secret:
+              secretName: create-dumps-{{ .Values.deployEnv }}
```
secret reference needs to align with reworked secret creation

BrendanGalloway commented on helm-chart/templates/cronjobs/mongobackup.yaml:
```diff
@@ -0,0 +1,106 @@
+              defaultMode: 0744
+          - name: upload-dumps-script
+            secret:
+              secretName: upload-dumps-{{ .Values.deployEnv }}
```
secret reference needs to align with reworked secret creation

BrendanGalloway commented on helm-chart/templates/secrets.yaml:
```diff
@@ -0,0 +1,78 @@
+  gcloud-service-accnt.json: {{ .Values.web.secrets.backupManager.serviceAccount | b64enc }}
+{{- end }}
+---
+{{- if or .Values.web.secrets.logging.prod .Values.web.secrets.logging.dev }}
```
why would this be environmentally keyed?  Every env would have its own helm deployment, describing web.secrets.logging for that environment

BrendanGalloway commented on helm-chart/templates/secrets.yaml:
```diff
@@ -0,0 +1,78 @@
+    deployEnv: "{{ .Values.deployEnv }}"
+type: Opaque
+data:
+  local_settings.py: {{ .Values.secrets.localSettings | b64enc }}
```
Is this transformable into a configmap file with env var lookups?

BrendanGalloway commented on helm-chart/templates/web.yaml:
```diff
@@ -0,0 +1,216 @@
+          subPath: gunicorn.conf.py
+          readOnly: true
+      volumes:
+        - name:  local-settings
+          secret:
+            secretName: local-settings-{{ .Values.deployEnv }}
+        - name: client-secret
+          secret:
+            secretName: google-client-secret-{{ .Values.deployEnv }}
+        - name: backup-manager-secret  # used to access google cloud
+          secret:
+            secretName: backup-manager-secret-{{ .Values.deployEnv }}
+        - name: logging-secret
+          secret:
+            secretName: logging-secret-{{ .Values.deployEnv }}
+        - name: varnish-secret
+          secret:
+            secretName: varnish-secret-{{ .Values.deployEnv }}
+        - name: school-lookup-data
+          secret:
+            secretName: school-lookup-data-{{ .Values.deployEnv }}
+        - name: gunicorn-config
+          configMap:
+            Name: gunicorn-config-{{ .Values.deployEnv }}
```
All secret references should contain a toggle to either a reference to an existing secret, or the envNamed secret generated if data is supplied.

BrendanGalloway commented on helm-chart/templates/web.yaml:
```diff
@@ -0,0 +1,216 @@
+          - name: VARNISH_HOST
+            value: "varnish-{{ .Values.deployEnv }}-{{ .Release.Revision }}"
+          - name: HELM_REVISION
+            value: "{{ .Release.Revision }}"
```
There needs to be an envFrom here to allow injection of the values required to process the templated configmaps

BrendanGalloway committed:
feat: do not allow multiple web pods to run on the same node

parth-gohil committed:
fix(helm): Handling secret references and creation.
This is done based on data provided for the secret via the data fields or references to an
existing secret in the cluster by providing it's name.

parth-gohil committed:
fix(helm): Better handle slack webhook URL secret.

BrendanGalloway commented on helm-chart/templates/cronjobs/daily_email_notifications.yaml:
```diff
@@ -0,0 +1,42 @@
+          volumes:
+          - name: local-settings
+            secret:
+              secretName: {{- if .Values.secrets.localSettings.ref }} {{- .Values.secrets.localSettings.ref }} {{- else }} local-settings-{{ .Values.deployEnv }} {{- end }}
```
This is still a bit off - currently this works if all of localsettings is passed in as the secret.  What we ideally want is a version of local settings with more env lookups mounted here as a configMap volume, and the switched secret mounted in envFrom

BrendanGalloway committed:
feat: example _helper.tpl functionality

parth-gohil committed:
fix(helm): Managing secret name selection via templates.

parth-gohil committed:
fix(helm): Managing secret name selection via templates for web pod.

parth-gohil committed:
fix(helm): Local settings to configmap + bug fixes
- Converted local settings to configmap
- Moved all the secrets to secrets and handled secrets conditionally.
- Several bug fixes to previous commits that were breaking the chart.
- Fixes to the template file and template application.

moshloop commented on helm-chart/values.yaml:
```diff
@@ -164,16 +164,109 @@ monitor:
+    ref: 
+    # Add the value of these variables in the data: field for local use only. 
+    # Also, it is understood that if you're adding one value you're adding variable values 
+    DATABASES_HOST:
```
Why do we have a data element ? Why not just go DATABASE_HOST: test

parth-gohil replied:
Earlier it was to differentiate between certain secrets that were individually defined. I stuck with the same convention for easy readability by devs in future.

moshloop replied:
Except that this isn't a standard convention - I have not seen other helm charts use this.

parth-gohil replied:
Noted. I'll remove the `data:` element.

moshloop commented on helm-chart/values.yaml:
```diff
@@ -164,16 +164,109 @@ monitor:
+  APSCHEDULER_NAME: "apscheduler-${ENV_NAME}"
+  SEARCH_ADMIN: "http://elasticsearch-data:9200"
+  TURN_SERVER: ''
+  USE_CLOUDFLARE: true
```
This should probably default to false

parth-gohil replied:
The value is true for sandbox, pre-prod and prod all use the value true by default :|
Should I still default to false?

moshloop commented on helm-chart/templates/configMaps.yaml:
```diff
@@ -349,3 +349,320 @@ data:
+        }
+    }
+
+    # TODO: Make the logging format more configurable based on specific objects being defined.
```
I think we can put the entire JSON value into configmap key

parth-gohil replied:
The logging objects change based on environment and there are several others that need to be added. When discussed with Brendan he said for a first pass we can go ahead with the prod configuration as default and then make others configurable within the python script based on environment variables set.

moshloop commented on helm-chart/templates/cronjobs/mongobackup.yaml:
```diff
@@ -0,0 +1,256 @@
+{{- if eq .Values.deployEnv "prod" }}
```
I am not sure that this or the nationbuilder stuff should be included in the public char

parth-gohil committed:
fix(helm): removes data: key from most secrets.

parth-gohil committed:
chore(helm): Sets the USE_CLOUDFLARE to false.

parth-gohil committed:
fix: Adds missing values.

parth-gohil committed:
fix: Updates the correct key.

parth-gohil committed:
chore: Patched changes from k8s repo.

BrendanGalloway commented on helm-chart/values.yaml:
```diff
@@ -0,0 +1,234 @@
+      serviceName: nginx
+      port: 80
+    - host: "{{ .Values.deployEnv }}.sefaria.org.il"
+    - host: sefaria.org
+    - host: sefaria.org.il
+    - host: www.sefaria.org
+    - host: www.sefaria.org.il
+    - host: rollout.sefaria.org
```
This should probably not be the default?

parth-gohil replied:
Can I just remove the prod links?

parth-gohil replied:
I've set the defaults for sandbox here.

BrendanGalloway commented on helm-chart/templates/secrets.yaml:
```diff
@@ -0,0 +1,125 @@
+data:
+  {{- with .Values.secrets.localSettings }}
+  DATABASES_HOST: {{ .DATABASES_HOST | b64enc }}
+  DATABASES_PASS: {{ .DATABASES_PASS | b64enc }}
+  DATABASES_USER: {{ .DATABASES_USER | b64enc }}
+  DATABASES_PORT: {{ .DATABASES_PORT | b64enc }}
+  GOOGLE_OAUTH2_CLIENT_ID: {{ .GOOGLE_OAUTH2_CLIENT_ID | b64enc }}
+  GOOGLE_OAUTH2_CLIENT_SECRET: {{ .GOOGLE_OAUTH2_CLIENT_SECRET | b64enc }}
+  SECRET_KEY: {{ .SECRET_KEY | b64enc }}
+  MANDRILL_API_KEY: {{ .MANDRILL_API_KEY | b64enc }}
+  SEFARIA_DB: {{ .SEFARIA_DB | b64enc }}
+  SEFARIA_DB_USER: {{ .SEFARIA_DB_USER | b64enc }}
+  SEFARIA_DB_PASSWORD: {{ .SEFARIA_DB_PASSWORD | b64enc }}
+  SEARCH_ADMIN_USER: {{ .SEARCH_ADMIN_USER | b64enc }}
+  SEARCH_ADMIN_PW: {{ .SEARCH_ADMIN_PW | b64enc }}
+  SEARCH_ADMIN_K8S: {{ .SEARCH_ADMIN_K8S | b64enc }}
+  TURN_SECRET: {{ .TURN_SECRET | b64enc }}
+  TURN_USER: {{ .TURN_USER | b64enc }}
+  SEFARIA_BOT_API_KEY: {{ .SEFARIA_BOT_API_KEY | b64enc }}
+  CLOUDFLARE_ZONE: {{ .CLOUDFLARE_ZONE | b64enc }}
+  CLOUDFLARE_EMAIL: {{ .CLOUDFLARE_EMAIL | b64enc }}
+  CLOUDFLARE_TOKEN: {{ .CLOUDFLARE_TOKEN | b64enc }}
+  GOOGLE_TAG_MANAGER_CODE: {{ .GOOGLE_TAG_MANAGER_CODE | b64enc }}
+  GOOGLE_ANALYTICS_CODE: {{ .GOOGLE_ANALYTICS_CODE | b64enc }}
+  GOOGLE_MAPS_API_KEY: {{ .GOOGLE_MAPS_API_KEY | b64enc }}
+  MIXPANEL_CODE: {{ .MIXPANEL_CODE | b64enc }}
+  HOTJAR_ID: {{ .HOTJAR_ID | b64enc }}
+  AWS_ACCESS_KEY: {{ .AWS_ACCESS_KEY | b64enc }}
+  AWS_SECRET_KEY: {{ .AWS_SECRET_KEY | b64enc }}
+  S3_BUCKET: {{ .S3_BUCKET | b64enc }}
+  NATIONBUILDER_TOKEN: {{ .NATIONBUILDER_TOKEN | b64enc }}
+  NATIONBUILDER_CLIENT_ID: {{ .NATIONBUILDER_CLIENT_ID | b64enc }}
+  NATIONBUILDER_CLIENT_SECRET: {{ .NATIONBUILDER_CLIENT_SECRET | b64enc }}
+  MAILCHIMP_API_KEY: {{ .MAILCHIMP_API_KEY | b64enc }}
+  MAILCHIMP_ANNOUNCE_ID: {{ .MAILCHIMP_ANNOUNCE_ID | b64enc }}
+  MAILCHIMP_WEBHOOK_KEY: {{ .MAILCHIMP_WEBHOOK_KEY | b64enc }}
+  RECAPTCHA_PUBLIC_KEY: {{ .RECAPTCHA_PUBLIC_KEY | b64enc }}
+  RECAPTCHA_PRIVATE_KEY: {{ .RECAPTCHA_PRIVATE_KEY | b64enc }}
+  SIMPLE_JWT_SIGNING_KEY: {{ .SIMPLE_JWT_SIGNING_KEY | b64enc }}
+  MOBILE_APP_KEY: {{ .MOBILE_APP_KEY | b64enc }}
```
I think for ease of maintenance and setting up environments where a subset of these values differ, it might be a good idea to split these out into a set of secrets, each of which covers a different configuration aspect.  That way someone would be able to update just AWS or database via the values file without also having to also populate all the other fields via values.  Or we could implement a default-override fallthrough set of imports.  @moshloop what do you think?

parth-gohil replied:
Waiting on this.

BrendanGalloway commented on helm-chart/templates/ingress.yaml:
```diff
@@ -0,0 +1,98 @@
+    - secretName: {{ template "secrets.originTls" . }}
+  rules:
+    - host: democracy.sefaria.org
+      http:
+        paths:
+          - path: /*
+            backend:
+              serviceName: nginx-{{ .Values.deployEnv }}
+              servicePort: 80
+---
+{{- if and .Values.ingress.secrets.originTls.cert.data .Values.ingress.secrets.originTls.key.data }}
+apiVersion: v1
+kind: Secret
+metadata:
+  name: origin-tls-{{ .Values.deployEnv }}
+type: kubernetes.io/tls
+data:
+  tls.crt: {{ .Values.ingress.secrets.originTls.cert.data | b64enc }}
+  tls.key: {{ .Values.ingress.secrets.originTls.key.data | b64enc }}
+{{- end }}
+{{- end }}
+---
+{{ if eq .Values.sandbox "true" }}
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: {{ .Values.deployEnv }} 
+  annotations:
+    kubernetes.io/ingress.class: nginx-ingress
+spec:
+  tls:
+    - secretName: cauldron-wildcard-letsencrypt       #wildcard cert managed via infrastructure
+  rules:
+  - host: "{{ .Values.deployEnv }}.cauldron.sefaria.org"
+    http:
+      paths:
+        - backend:
+            serviceName: nginx-{{ .Values.deployEnv }}
+            servicePort: 80
+{{ end }}
```
It should be possible to consolidate this down to a single ingress template, now that the values has the ability to specify the host-backend mapping.  The only different between democracy and prod is what they point at, and then sandboxes additionally need to use the nginx ingress annotation

parth-gohil replied:
Let me see what I can do for this.

parth-gohil replied:
I've refactored this. Do take a look at this once to see if the previous logic is preserved.

parth-gohil committed:
chore(helm): Refactored the ingress.
Also, updates the default ingress values.

BrendanGalloway commented on helm-chart/templates/ingress.yaml:
```diff
@@ -0,0 +1,47 @@
+    kubernetes.io/ingress.global-static-ip-name: "sefaria-{{ .Values.deployEnv }}"
+    {{- end }}
+spec:
+  tls:
+    - secretName: {{ template "secrets.originTls" . }}
+    - secretName: {{ template "secrets.originIlTls" . }}
```
This section also switches on sandbox - all sandboxes use the wildcard cert.  Unless you are managing that the values file?  The IlTls reference is also only required in prod.

parth-gohil replied:
Ya... missed that part. 

Will make the second secret to be conditional but I am handling the secret name from the Values file so we can let the first secret name be pulled from the values file. I've left a comment in the values file for the same.

parth-gohil committed:
fix(helm): Conditionally using ingress secret.

parth-gohil committed:
fix(helm): Bug fix for templates and values file.
- Removes unnecessary leading spaces in 2 templates.
- Changes the hierarchy of data keys in values file.
- Minor bug fix in the nodejs file.
- Empty space cleanup in monitor file.
- Fixed the scope of certain template variables in ingress file.

parth-gohil committed:
fix(helm): Updates ingress to networking/v1 values

parth-gohil committed:
fix(helm): Sets the values default for secrets.
Sets up the chart to use existing secrets in the cluster as a default.

parth-gohil committed:
feat(helm): Adds mongo restore job.

parth-gohil committed:
fix(helm): Ingress path removal & setting pathType

parth-gohil committed:
fix(helm): Removes DOMAIN_LANGUAGE defaults.
This was causing the 500 statuses to /api/background-data on nginx pod

parth-gohil committed:
fix(helm): Helmizes missed APSCHEDULER_NAME value.
This is what was causing the monitor to fail repeatedly.

parth-gohil committed:
feat(helm): No PodDisruptionBudget in sandbox.
